pallets/subtensor/src/utils.rs

```diff
@@ -598,4 +599,40 @@ impl<T: Config> Pallet<T> {
     pub fn is_subnet_owner(address: &T::AccountId) -> bool {
         SubnetOwner::<T>::iter_values().any(|owner| *address == owner)
     }
+
+    pub fn log2(x: f64) -> f64 {
+        let mut y = x;
+        let mut result = 0.0;
+        while y < 1.0 {
+            y *= 2.0;
+            result -= 1.0;
+        }
+        while y >= 2.0 {
+            y *= 0.5;
+            result += 1.0;
+        }
+        result
+    }
+    pub fn powf(base: f64, exponent: f64) -> f64 {
+        if exponent == 0.0 {
+            return 1.0;
+        } else if exponent < 0.0 {
+            return 1.0 / Self::powf(base, -exponent);
+        }
+
+        let mut result = 1.0;
+        let mut exp = exponent as i64;
+        let mut b = base;
+        while exp > 0 {
+            if exp % 2 == 1 {
+                result *= b;
+            }
+            b *= b;
+            exp /= 2;
+        }
+        result
+    }
+    pub fn floor(x: f64) -> f64 {
+        x as i64 as f64
+    }
```

Comment:
so I should have thought about this before I helped you get f64 stuff working, but probably a bad idea to have anything on-chain rely on an f64 calculation

I would recommend instead using the `rust_decimal` crate with default features disabled, which disables the `std` feature allowing for `no_std` support which will let you use it in pallets etc. This way we will get predictable / deterministic outputs and eliminate floating point weirdness being a source of inconsistency